compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -468,6 +460,52 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             }
         })
     }
+
+    fn check_offset_align(
+        &self,
+        offset: u64,
+        align: Align,
+        check: CheckAlignment,
+    ) -> InterpResult<'tcx> {
+        if offset % align.bytes() == 0 {
+            Ok(())
+        } else {
+            // The biggest power of two through which `offset` is divisible.
+            let offset_pow2 = 1 << offset.trailing_zeros();
+            self.check_alignment(Align::from_bytes(offset_pow2).unwrap(), align, check)
+        }
+    }
+
+    fn check_alignment(
+        &self,
+        has: Align,
+        required: Align,
+        check: CheckAlignment,
+    ) -> InterpResult<'tcx, ()> {
+        match check {
+            CheckAlignment::Error => {
+                throw_ub!(AlignmentCheckFailed { has, required })
+            }
+            CheckAlignment::No => {}
+            CheckAlignment::FutureIncompat => self.tcx.struct_span_lint_hir(
+                INVALID_ALIGNMENT,
+                self.stack().iter().find_map(|frame| frame.lint_root()).unwrap_or(CRATE_HIR_ID),
+                self.cur_span(),
+                UndefinedBehaviorInfo::AlignmentCheckFailed { has, required }.to_string(),
+                |db| {
+                    let mut stacktrace = self.generate_stacktrace();
+                    // Filter out `requires_caller_location` frames.
+                    stacktrace
+                        .retain(|frame| !frame.instance.def.requires_caller_location(*self.tcx));
+                    for frame in stacktrace {
+                        db.span_label(frame.span, format!("inside `{}`", frame.instance));
+                    }
+                    db
+                },
+            ),
```

Comment:
Is it possible to move this logic out of the interpreter, into the const_eval machine instance? That would also let us share the code with the backtrace generation there, rather than having yet another copy of that logic.

Reply:
> That would also let us share the code with the backtrace generation there, rather than having yet another copy of that logic.

doing that requires reverting making that logic always error and allow us to choose lint vs error again 😆 I did consider it

Reply:
I think we can do it in a way that is less messy than what we had before... instead of having report_decorated do both of these, we can factor the logic of capturing a stacktrace from the interpreter and adding the spans to a diagnostic (probably filtering should just be part of the latter) out so that `ConstEvalErr` and a new `const_eval_lint` function can both call that shared logic.